src/bio.c

```diff
@@ -61,24 +63,47 @@
 #include "server.h"
 #include "bio.h"
 
-static pthread_t bio_threads[BIO_NUM_OPS];
-static pthread_mutex_t bio_mutex[BIO_NUM_OPS];
-static pthread_cond_t bio_newjob_cond[BIO_NUM_OPS];
-static list *bio_jobs[BIO_NUM_OPS];
+static char* bio_worker_title[] = {
+    "bio_close_file",
+    "bio_aof",
+    "bio_lazy_free",
+};
+
+#define BIO_WORKER_NUM (sizeof(bio_worker_title) / sizeof(*bio_worker_title))
+
+static unsigned int bio_job_to_worker[] = {
+    [BIO_CLOSE_FILE] = 0,
+    [BIO_AOF_FSYNC] = 1,
+    [BIO_CLOSE_AOF] = 1,
+    [BIO_LAZY_FREE] = 2,
+};
+
+static pthread_t bio_threads[BIO_WORKER_NUM];
+static pthread_mutex_t bio_mutex[BIO_WORKER_NUM];
+static pthread_cond_t bio_newjob_cond[BIO_WORKER_NUM];
+static list *bio_jobs[BIO_WORKER_NUM];
+static unsigned long bio_jobs_counter[BIO_NUM_OPS] = {0};
 
 /* This structure represents a background Job. It is only used locally to this
  * file as the API does not expose the internals at all. */
 typedef union bio_job {
+    struct {
+        int type; /* Job-type tag. This needs to appear as the first element in all union members. */
```

Comment:
Did you consider the common pattern of have a struct with a `type` member followed by the union with all members? It might be a bit cleaner.

Reply:
i think we considered it, it caused more blame log damage and we opted against it.
@slavak correct me if i'm wrong.

Reply:
Mainly because `free_args[]` can't be used in union.
`flexible array member in a struct with no named member`

Reply:
```c
typedef struct bio_job {
    int type; /* Job-type tag. This needs to appear as the first element in all union members. */

    union {
        /* Job specific arguments.*/
        struct {
            int type;
            int fd; /* Fd for file based background jobs */
            long long offset; /* A job-specific offset, if applicable */
            unsigned need_fsync:1; /* A flag to indicate that a fsync is required before
                                    * the file is closed. */
            unsigned need_reclaim_cache:1; /* A flag to indicate that reclaim cache is required before
                                    * the file is closed. */
        } fd_args;

        struct {
            int type;
            void *free_args[]; /* List of arguments to be passed to the free function */
        } free_args;
    };
    
} bio_job;
```
compile warning: `warning: invalid use of structure with flexible array member [-Wpedantic]`